lib/commons/dom/get-element-stack.js

```diff
@@ -0,0 +1,489 @@
+ */
+function createGrid() {
+	const rootGrid = {
+		container: null,
+		cells: []
+	};
+
+	axe.utils
+		.querySelectorAll(axe._tree[0], '*')
+		.filter(vNode => vNode.actualNode.parentElement !== document.head)
+		.forEach(vNode => {
+			if (vNode.actualNode.nodeType !== 1) {
+				return;
+			}
+
+			vNode._stackingOrder = getStackingOrder(vNode);
+
+			// filter out any elements with 0 width or height
+			// (we don't do this before so we can calculate stacking context
+			// of parents with 0 width/height)
+			const rect = vNode.boundingClientRect;
+			if (rect.width === 0 || rect.height === 0) {
+				return;
+			}
+
+			const scrollRegionParent = findScrollRegionParent(vNode);
+			const grid = scrollRegionParent ? scrollRegionParent._subGrid : rootGrid;
+
+			if (isScrollRegion(vNode)) {
+				const subGrid = {
+					container: vNode,
+					cells: []
+				};
+				vNode._subGrid = subGrid;
+			}
+
+			addNodeToGrid(grid, vNode);
+		});
+}
+
+/**
+ * Return all elements that are at the center point of the passed in virtual node.
+ * @method getElementStack
+ * @memberof axe.commons.dom
+ * @param {VirtualNode} vNode
+ * @param {Boolean} recursed If the function has been called recursively
+ * @return {VirtualNode[]}
+ */
+dom.getElementStack = function(vNode, recursed) {
+	if (!axe._cache.get('gridCreated')) {
+		createGrid();
+		axe._cache.set('gridCreated', true);
+	}
+
+	const grid = vNode._grid;
+
+	if (!grid) {
+		return [];
+	}
+
+	const boundingRect = vNode.boundingClientRect;
+
+	// use center point of rect
+	let x = getDomPosition(boundingRect, 'x') + boundingRect.width / 2;
+	let y = getDomPosition(boundingRect, 'y') + boundingRect.height / 2;
+
+	// NOTE: there is a very rare edge case in Chrome vs Firefox that can
+	// return different results of `document.elementsFromPoint`. If the center
+	// point of the element is <1px outside of another elements bounding rect,
+	// Chrome appears to round the number up and return the element while Firefox
+	// keeps the number as is and won't return the element. In this case, we
+	// went with pixel perfect collision rather than rounding
+	const row = Math.floor(y / gridSize);
+	const col = Math.floor(x / gridSize);
+	let stack = grid.cells[row][col].filter(vNode => {
+		const rects = vNode.clientRects;
+		const nodeName = vNode.actualNode.nodeName.toLowerCase();
+		return rects.find(rect => {
+			let pointX = x;
+			let pointY = y;
+
+			let rectWidth = rect.width;
+			let rectHeight = rect.height;
+			let rectX = getDomPosition(rect, 'x');
+			let rectY = getDomPosition(rect, 'y');
+
+			const transform =
+				vNode.getComputedStylePropertyValue('-webkit-transform') ||
+				vNode.getComputedStylePropertyValue('-ms-transform') ||
+				vNode.getComputedStylePropertyValue('transform') ||
+				'none';
+
+			// if rect is rotated, rotate point around center of rect and then
+			// perform collision detection as though the rect was not rotated
+			if (
```

Comment:
Similarly, if rect is clipped, treat it as not clipped. It could "stop" the stack or allow the stack to continue depending on whether it's a masking or keyhole technique. We shouldn't try to handle this case, just note that it's an area we can't address without a canvas to plot on.